File: crates/markdown-contract-engine/src/registry.rs
//! The finding-id → default `FindingLevel` registry plus the `Ctx` factory.
//!
//! `level` is **contract data**, not a call-site choice (D-0001): a finding id has one
//! default severity wherever it fires. The tables mirror the TS engine's `registry.ts`
//! verbatim — the structure, content, text, and rule planes' defaults — and an
//! unregistered id defaults to `"error"`.
//!
//! [`Ctx`] is the rule-author finding factory: `ctx.finding(spec)` stamps the document
//! `path`, fills `level` from the registry when the caller does not supply one, and
//! carries `pos` only when supplied (omitted for absence findings).

use std::collections::HashMap;

use crate::finding::{Finding, FindingLevel, SourcePos};

/// The default severity for every structure-plane finding id (D-0003).
pub const STRUCTURE_LEVELS: &[(&str, FindingLevel)] = &[
    ("structure/section-missing", FindingLevel::Error),
    ("structure/section-order", FindingLevel::Error),
    ("structure/duplicate-section", FindingLevel::Error),
    ("structure/key-collision", FindingLevel::Error),
    ("structure/anchor-missing", FindingLevel::Error),
    ("structure/block-missing", FindingLevel::Error),
    ("structure/block-kind", FindingLevel::Error),
    ("structure/gap-count", FindingLevel::Error),
    ("structure/repeat-count", FindingLevel::Error),
    ("structure/heading-depth-jump", FindingLevel::Warn),
];

/// The default severity for every content- and frontmatter-plane finding id (D-0004 / D-0001).
/// The content plane itself lands next phase; the ids are registered now so severity
/// resolution is complete from day one.
pub const CONTENT_LEVELS: &[(&str, FindingLevel)] = &[
    ("content/table/column-missing", FindingLevel::Error),
    ("content/table/column-extra", FindingLevel::Error),
    ("content/table/min-rows", FindingLevel::Error),
    ("content/table/cell", FindingLevel::Error),
    ("content/list/item-kind", FindingLevel::Error),
    ("content/list/min-items", FindingLevel::Error),
    ("content/code/lang", FindingLevel::Error),
    ("content/max-words", FindingLevel::Error),
    ("frontmatter/enum", FindingLevel::Error),
    ("frontmatter/unknown-key", FindingLevel::Error),
    ("frontmatter/type", FindingLevel::Error),
    ("frontmatter/required", FindingLevel::Error),
    ("frontmatter/refine", FindingLevel::Error),
];

/// Default severities for the declarative text plane (`text/*` — D-0011).
pub const TEXT_LEVELS: &[(&str, FindingLevel)] = &[
    ("text/requires", FindingLevel::Error),
    ("text/forbids", FindingLevel::Error),
    ("text/count", FindingLevel::Error),
];

/// Default severities for the rule plane — the contract-chosen namespaces the corpus
/// contracts use. An unregistered rule id still defaults to `"error"`, so this table is
/// the documented, not the load-bearing, source.
pub const RULE_LEVELS: &[(&str, FindingLevel)] = &[
    ("task/post-mortem-when-worked", FindingLevel::Error),
    ("task/completion-note-when-closed", FindingLevel::Error),
    ("summary/mentions-outcome", FindingLevel::Error),
    ("summary/names-contract", FindingLevel::Warn),
];

/// The id → default-level registry, seeded with every plane's defaults.
#[derive(Debug, Clone)]
pub struct Registry {
    levels: HashMap<&'static str, FindingLevel>,
}

impl Default for Registry {
    fn default() -> Self {
        let levels = STRUCTURE_LEVELS
            .iter()
            .chain(CONTENT_LEVELS)
            .chain(TEXT_LEVELS)
            .chain(RULE_LEVELS)
            .copied()
            .collect();
        Self { levels }
    }
}

impl Registry {
    /// The default level for `id`; an unregistered id defaults to `"error"`.
    pub fn level_for(&self, id: &str) -> FindingLevel {
        self.levels.get(id).copied().unwrap_or(FindingLevel::Error)
    }
}

/// The input to [`Ctx::finding`] — id and message required, level/pos optional.
#[derive(Debug, Clone)]
pub struct FindingSpec {
    pub id: String,
    pub message: String,
    pub level: Option<FindingLevel>,
    pub pos: Option<SourcePos>,
}

impl FindingSpec {
    pub fn new(id: impl Into<String>, message: impl Into<String>) -> Self {
        Self { id: id.into(), message: message.into(), level: None, pos: None }
    }

    /// Override the registry default level.
    pub fn level(mut self, level: FindingLevel) -> Self {
        self.level = Some(level);
        self
    }

    /// Pin the finding to a source position (omit for whole-document absence findings).
    pub fn pos(mut self, pos: SourcePos) -> Self {
        self.pos = Some(pos);
        self
    }
}

/// The rule author's finding factory — the engine fills `path` and the id's default level.
pub struct Ctx<'a> {
    pub path: &'a str,
    registry: &'a Registry,
}

impl<'a> Ctx<'a> {
    pub fn new(path: &'a str, registry: &'a Registry) -> Self {
        Self { path, registry }
    }

    /// Build a [`Finding`]: stamps `path`, fills `level` from the registry (defaulting to
    /// `"error"` for an unregistered id), carries `pos` only when supplied.
    pub fn finding(&self, spec: FindingSpec) -> Finding {
        let level = spec.level.unwrap_or_else(|| self.registry.level_for(&spec.id));
        Finding { id: spec.id, level, path: self.path.to_string(), pos: spec.pos, message: spec.message, fix: None }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // Contract first: a structure id resolves to its registered default; heading-depth-jump
    // is the one warn-level structure default; an unregistered id falls back to error.
    #[test]
    fn registry_defaults_match_the_ts_tables() {
        let r = Registry::default();
        assert_eq!(r.level_for("structure/section-missing"), FindingLevel::Error);
        assert_eq!(r.level_for("structure/heading-depth-jump"), FindingLevel::Warn);
        assert_eq!(r.level_for("summary/names-contract"), FindingLevel::Warn);
        assert_eq!(r.level_for("text/requires"), FindingLevel::Error);
        assert_eq!(r.level_for("totally/unregistered"), FindingLevel::Error);
    }

    #[test]
    fn ctx_finding_stamps_path_and_default_level() {
        let r = Registry::default();
        let ctx = Ctx::new("docs/note.md", &r);
        let f = ctx.finding(FindingSpec::new("structure/heading-depth-jump", "skip").pos(SourcePos::at(5, 1)));
        assert_eq!(f.level, FindingLevel::Warn);
        assert_eq!(f.path, "docs/note.md");
        assert_eq!(f.pos, Some(SourcePos::at(5, 1)));
    }

    #[test]
    fn explicit_level_overrides_the_registry() {
        let r = Registry::default();
        let ctx = Ctx::new("a.md", &r);
        let f = ctx.finding(FindingSpec::new("structure/section-missing", "m").level(FindingLevel::Report));
        assert_eq!(f.level, FindingLevel::Report);
    }

    #[test]
    fn pos_is_absent_unless_supplied() {
        let r = Registry::default();
        let ctx = Ctx::new("a.md", &r);
        assert_eq!(ctx.finding(FindingSpec::new("x/y", "m")).pos, None);
    }
}

File: packages/core/src/core/registry.ts
/**
 * The finding-id → default `FindingLevel` registry plus the `Ctx` factory (`makeCtx`).
 *
 * `level` is **contract data**, not a call-site choice (D-0001): a finding id has one
 * default severity wherever it fires. This registry holds the structure- and content-plane
 * defaults; the rule (`rule/*` / `docRule`) defaults are added when that plane lands (T-3NC8).
 *
 * `makeCtx(path, registry)` returns the `Ctx` a rule body uses: its `finding(...)` stamps
 * the document `path` and fills `level` from the registry when the caller does not supply
 * one (A4). Engine-internal structure findings reuse the same registry through this Ctx so
 * a single source defines every default.
 */
import type { Ctx, Finding, FindingLevel, SourcePos } from "./types.js";

/** The default severity for every structure-plane finding id this plane emits (D-0003). */
export const STRUCTURE_LEVELS: Record<string, FindingLevel> = {
  "structure/section-missing": "error",
  "structure/section-order": "error",
  "structure/duplicate-section": "error",
  "structure/key-collision": "error",
  "structure/anchor-missing": "error",
  "structure/block-missing": "error",
  "structure/block-kind": "error",
  "structure/gap-count": "error",
  "structure/repeat-count": "error",
  "structure/heading-depth-jump": "warn",
};

/**
 * The default severity for every content- and frontmatter-plane finding id (D-0004 / D-0001).
 * Content checks default to `error` (a data-shape violation blocks the typed model); the
 * frontmatter Zod findings are likewise `error`.
 */
const CONTENT_LEVELS: Record<string, FindingLevel> = {
  // table
  "content/table/column-missing": "error",
  "content/table/column-extra": "error",
  "content/table/min-rows": "error",
  "content/table/cell": "error",
  // list
  "content/list/item-kind": "error",
  "content/list/min-items": "error",
  // code
  "content/code/lang": "error",
  // paragraph
  "content/max-words": "error",
  // frontmatter (Zod over the YAML)
  "frontmatter/enum": "error",
  "frontmatter/unknown-key": "error",
  "frontmatter/type": "error",
  "frontmatter/required": "error",
  // a `.refine()` cross-field predicate surfaces as a Zod `custom` issue (D-0001 E1)
  "frontmatter/refine": "error",
};

/**
 * Default severities for the rule plane (`rule/*` — and the contract-chosen namespaces a
 * `rule` / `docRule` may mint, e.g. `task/...`, `summary/...`). `level` is contract data
 * (D-0001 A4): a rule body names the problem and the engine fills the default level. A rule
 * that supplies its own `level` overrides this; an unregistered rule id still defaults to
 * `"error"` via {@link makeCtx}, so this table is the documented, not the load-bearing, source.
 */
const RULE_LEVELS: Record<string, FindingLevel> = {
  // cross-plane docRule ids the corpus contracts use
  "task/post-mortem-when-worked": "error",
  "task/completion-note-when-closed": "error",
  // node-local rule ids the corpus contracts use
  "summary/mentions-outcome": "error",
  "summary/names-contract": "warn",
};

/**
 * Default severities for the text plane (`text/*` — D-0011). Every declarative text constraint
 * emits through these three ids: `text/requires` (a required phrase is missing), `text/forbids`
 * (a forbidden phrase is present), and `text/count` (a `min` / `max` occurrence bound is
 * violated). All default to `error`, overridable per entry via the spec's `level`.
 */
const TEXT_LEVELS: Record<string, FindingLevel> = {
  "text/requires": "error",
  "text/forbids": "error",
  "text/count": "error",
};

/** The id → default-level registry. T-3NC8 extends it with rule ids. */
type LevelRegistry = Record<string, FindingLevel>;

/** A fresh registry seeded with the structure-, content-, text-, and rule-plane defaults. */
export function defaultRegistry(): LevelRegistry {
  return { ...STRUCTURE_LEVELS, ...CONTENT_LEVELS, ...TEXT_LEVELS, ...RULE_LEVELS };
}

/**
 * Build the rule-author finding factory. `finding({id, message, level?, pos?})` stamps
 * the document `path`, fills `level` from the registry (defaulting to `"error"` for an
 * unregistered id), and carries `pos` only when supplied (omitted for absence findings).
 */
export function makeCtx(path: string, registry: LevelRegistry): Ctx {
  return {
    path,
    finding(f: { id: string; message: string; level?: FindingLevel; pos?: SourcePos }): Finding {
      const level: FindingLevel = f.level ?? registry[f.id] ?? "error";
      const out: Finding = { id: f.id, level, path, message: f.message };
      if (f.pos !== undefined) out.pos = f.pos;
      return out;
    },
  };
}

/**
 * Decorate a `Ctx` so every finding it mints carries `hint` — the nearest enclosing authored
 * `description` in scope at the mint site (D-0020). The engine walks compute the EFFECTIVE hint
 * (own description ?? inherited) and wrap once per scope; a later wrap overrides an earlier one,
 * so the innermost description wins. `hint === undefined` returns `ctx` unchanged — a
 * description-free contract mints byte-identical findings (no `hint` key at all).
 */
export function withHint(ctx: Ctx, hint: string | undefined): Ctx {
  if (hint === undefined) return ctx;
  return {
    path: ctx.path,
    finding(f: { id: string; message: string; level?: FindingLevel; pos?: SourcePos }): Finding {
      const out = ctx.finding(f);
      out.hint = hint;
      return out;
    },
  };
}
